File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/D._Деревья_близнецы/solution.py
import os

LOCAL = os.environ.get('REMOTE_JUDGE', 'false') != 'true'

if LOCAL:
    class Node:
        def __init__(self, value, left=None, right=None):
            self.value = value
            self.right = right
            self.left = left


def solution(root1: 'Node', root2: 'Node') -> bool:
    """
    Time Complexity O(N)
    Memory Complexity O(H)"""
    if root1 == root2 == None: return  True
    elif (
        (root1 and not root2)
        or (not root1 and root2)
        or (root1.value != root2.value)
    ): return False
    return solution(root1.left, root2.left) and solution(root1.right, root2.right)


def test():
    node1 = Node(1, None, None)
    node2 = Node(2, None, None)
    node3 = Node(3, node1, node2)

    node4 = Node(1, None, None)
    node5 = Node(2, None, None)
    node6 = Node(3, node4, node5)

    assert solution(node3, node6)


if __name__ == '__main__':
    test()

File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/D._Деревья_близнецы/solution.cpp
#ifndef REMOTE_JUDGE
struct Node {
  int value;
  const Node* left = nullptr;
  const Node* right = nullptr;
  Node(int value, Node* left, Node* right) : value(value), left(left), right(right) {}
};
#endif

#ifdef REMOTE_JUDGE
#include "solution_tree.h"
#endif
#include <cmath>
#include <iostream>
#include <cassert>

using namespace std;


bool Solution(const Node* root1, const Node* root2) {
    /*
    Time Complexity O(N)
    Memory Complexity O(H)
    */
    if (!root1 && !root2) return true;
    if (
        (root1 && !root2)
        || (!root1 && root2)
        || (root1->value != root2->value)
    ) return false;
    return Solution(root1->left, root2->left) && Solution(root1->right, root2->right);
}

#ifndef REMOTE_JUDGE
void test() {

    Node node1({1, nullptr, nullptr});
    Node node2({2, nullptr, nullptr});
    Node node3({3, &node1, &node2});

    Node node4({1, nullptr, nullptr});
    Node node5({2, nullptr, nullptr});
    Node node6({3, &node4, &node5});
    assert(Solution(&node3, &node6));
}

int main() {
  test();
}
#endif